File: bench/cases/sort/sort.js
// sort: insertion sort (O(n^2)) of an LCG-filled Int32Array.
// LCG uses Math.imul + >>>0 to reproduce the exact uint32-wraparound sequence
// used by the Go and Zeus implementations.
// checksum = positional weighted sum mod MOD.
"use strict";

const N = 20000;
const MOD = 1000003;
const expected = 861863;

const arr = new Int32Array(N);
let x = 12345 >>> 0;
for (let i = 0; i < N; i++) {
  x = (Math.imul(x, 1103515245) + 12345) >>> 0;
  arr[i] = (x >>> 16) & 0xffff;
}

for (let i = 1; i < N; i++) {
  const key = arr[i];
  let j = i - 1;
  while (j >= 0 && arr[j] > key) {
    arr[j + 1] = arr[j];
    j--;
  }
  arr[j + 1] = key;
}

let sum = 0;
for (let i = 0; i < N; i++) {
  sum = (sum + (i + 1) * arr[i]) % MOD;
}
const checksum = sum;
if (checksum !== expected) {
  process.stderr.write(`MISMATCH got=${checksum} want=${expected}\n`);
  process.exit(1);
}
process.exit(0);

File: bench/cases/sort/sort.go
package main

// sort: insertion sort (O(n^2)) of an LCG-filled array — exercises array
// get/set and comparisons. The LCG uses uint32 wraparound so the exact same
// sequence is reproducible in Go, Zeus (i64 + mask) and JS (Math.imul).
// checksum = positional weighted sum mod MOD (detects both content and order).

import (
	"fmt"
	"os"
)

const n = 20000
const mod int64 = 1000003
const expected int64 = 861863

func main() {
	arr := make([]int32, n)
	var x uint32 = 12345
	for i := 0; i < n; i++ {
		x = x*1103515245 + 12345
		arr[i] = int32((x >> 16) & 0xffff)
	}

	for i := 1; i < n; i++ {
		key := arr[i]
		j := i - 1
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}

	var sum int64 = 0
	for i := 0; i < n; i++ {
		sum = (sum + int64(i+1)*int64(arr[i])) % mod
	}
	checksum := sum
	if expected < 0 {
		fmt.Fprintf(os.Stderr, "checksum=%d\n", checksum)
		os.Exit(0)
	}
	if checksum != expected {
		fmt.Fprintf(os.Stderr, "MISMATCH got=%d want=%d\n", checksum, expected)
		os.Exit(1)
	}
	os.Exit(0)
}
